engine/dlib/src/dlib/opaque_handle_container.h

```diff
@@ -0,0 +1,280 @@
+    uint16_t* m_ObjectVersions;
+    uint32_t  m_Capacity;
+    uint16_t  m_Version;
+
+    // Internal helper functions
+    inline T* GetByIndex(uint32_t i)
+    {
+        assert(i < m_Capacity);
+        return m_Objects[i];
+    }
+
+    uint32_t GetFirstEmptyIndex()
+    {
+        for (int i = 0; i < m_Capacity; ++i)
+        {
+            if (m_Objects[i] == 0)
+            {
+                return i;
+            }
+        }
+        return INVALID_OPAQUE_HANDLE;
+    }
+
+    dmOpaqueHandleContainer(const dmOpaqueHandleContainer<T>&) {}
+    void operator =(const dmOpaqueHandleContainer<T>&) {}
+    void operator ==(const dmOpaqueHandleContainer<T>&) {}
+};
+
+
+//////////////////////////////////////////////////////////////
+// Public API implementation
+//////////////////////////////////////////////////////////////
+
+template <typename T>
+dmOpaqueHandleContainer<T>::dmOpaqueHandleContainer()
+: m_Objects(0)
+, m_Capacity(0)
+, m_Version(0)
+{}
+
+template <typename T>
+dmOpaqueHandleContainer<T>::dmOpaqueHandleContainer(uint32_t inital_capacity)
+: dmOpaqueHandleContainer()
+{
+    Allocate(inital_capacity);
+}
+
+template <typename T>
+dmOpaqueHandleContainer<T>::~dmOpaqueHandleContainer()
+{
+    if (m_Objects != 0)
+    {
+        free(m_Objects);
+        free(m_ObjectVersions);
+    }
+}
+
+template <typename T>
+bool dmOpaqueHandleContainer<T>::Allocate(uint32_t count)
+{
+    if (m_Objects == 0x0)
+    {
+        m_Objects        = (T**) malloc(count * sizeof(sizeof(T*)));
+        m_ObjectVersions = (uint16_t*) malloc(count * sizeof(uint16_t));
+    }
+    else
+    {
+        uint32_t new_capacity = m_Capacity + count;
+        m_Objects             = (T**) realloc(m_Objects, new_capacity * sizeof(T*));
+        m_ObjectVersions      = (uint16_t*) realloc(m_ObjectVersions, new_capacity * sizeof(uint16_t));
+    }
+
+    memset(m_Objects + m_Capacity, 0, count * sizeof(T*));
+    memset(m_ObjectVersions + m_Capacity, 0, count * sizeof(uint16_t));
+
+    m_Capacity += count;
+
+    return m_Objects != 0x0 && m_ObjectVersions != 0x0;
+}
+
+template <typename T>
+T* dmOpaqueHandleContainer<T>::Get(HOpaqueHandle handle)
+{
+    if (handle == 0 || handle == INVALID_OPAQUE_HANDLE)
+    {
+        return 0;
+    }
+
+    uint32_t version = handle >> 16;
+    uint32_t index   = handle & 0xFFFF;
+    T* entry         = GetByIndex(index);
+
+    if (entry == 0 || m_ObjectVersions[index] != version)
+    {
+        return 0;
+    }
+
+    return entry;
+}
+
+template <typename T>
+HOpaqueHandle dmOpaqueHandleContainer<T>::Put(T* obj)
+{
+    uint32_t index = GetFirstEmptyIndex();
+    if (index == INVALID_OPAQUE_HANDLE)
+    {
+        return INVALID_OPAQUE_HANDLE;
+    }
+
+    m_Version++;
+    if (m_Version == 0)
+    {
+        // Set to 1 to avoid potentially producing a 0 handle
+        m_Version = 1;
+    }
+
+    m_ObjectVersions[index] = m_Version;
+    m_Objects[index]        = obj;
+
+    return m_Version << 16 | index;
```

Comment:
... and we return `(0xFFFF << 16) | 0xFFFF` which is equal to `INVALID_OPAQUE_HANDLE`, right?

I know I probably wrote this code, but we should probably fix this right now.
Also, it might be good to mention that the container cannot handle more than 65534 or 65535 handles.

Reply:
Yes, if both of them are 0xFFFF it will equate an invalid handle, so either we cap the version or the max size of the container so the max handle becomes 0xfffffffe or 0xfffeffff I suppose